pegasus/plotting/plot_library.py

```diff
@@ -1581,6 +1582,108 @@ def non_zero(g):
+        See `scipy linkage documentation`_ for details.
+    panel_size: ``Tuple[float, float]``, optional, default: ``(10, 6)``
+        The size (width, height) in inches of figure.
+    label_rotation: ``float``, optional, default: ``45``
+        Rotation angle of xtick labels.
+    orientation: ``str``, optional, default: ``top``
+        The direction to plot the dendrogram. Available options are: ``top``, ``bottom``, ``left``, ``right``. See `scipy dendrogram documentation`_ for explanation.
+    color_threshold``float``, optional, default: ``None``
+        Threshold for coloring clusters. See `scipy dendrogram documentation`_ for explanation.
+    return_fig: ``bool``, optional, default: ``False``
+        Return a ``Figure`` object if ``True``; return ``None`` otherwise.
+    dpi``float``, optional, default: ``300.0``
+        The resolution in dots per inch.
+
+    Returns
+    -------
+
+    ``Figure`` object
+        A ``matplotlib.figure.Figure`` object containing the dot plot if ``return_fig == True``
+
+    Examples
+    --------
+    >>> pg.dendrogram(data, rep='pca', groupby='leiden_labels')
+    >>> pg.dendrogram(data, genes=data.var_names, on_average=False)
+
+    .. _scipy linkage documentation: https://docs.scipy.org/doc/scipy/reference/generated/scipy.cluster.hierarchy.linkage.html
+    .. _scipy dendrogram documentation: https://docs.scipy.org/doc/scipy/reference/generated/scipy.cluster.hierarchy.dendrogram.html
+    """
+    if genes is None:
+        embed_df = pd.DataFrame(X_from_rep(data, rep))
+    else:
+        embed_df = pd.DataFrame(slicing(data[:, genes].X))
+    embed_df.set_index(
+        data.obs_names if groupby is None else data.obs[groupby], inplace=True
+    )
+
+    if on_average:
+        embed_df = embed_df.groupby(level=0, observed=True).mean()
+    if not isinstance(embed_df.index.dtype, pd.CategoricalDtype):
+        embed_df.index = embed_df.index.astype("category")
+    corr_df = pd.DataFrame(np.corrcoef(embed_df, rowvar=True), columns=embed_df.index, index=embed_df.index)  # Faster than pandas corr, but only Pearson metric
+    corr_mat = corr_df.values
+
+    from pegasus.tools.utils import calc_csi_matrix
+
+    csi_mat = calc_csi_matrix(corr_mat)
+    csi_df = pd.DataFrame(csi_mat, columns=corr_df.index, index=corr_df.index)
+
+    from scipy.cluster.hierarchy import linkage, dendrogram
+    from scipy.spatial.distance import squareform
+
+    dissim_df = 1 - csi_df
+    Z = linkage(squareform(dissim_df), method=linkage_method, optimal_ordering=True)
+    fig, ax = _get_subplot_layouts(panel_size=panel_size, dpi=dpi)
+    dendrogram(
+        Z,
+        ax=ax,
+        labels=dissim_df.index,
+        color_threshold=color_threshold,
+        orientation=orientation,
+    )
+    fig.tight_layout()
+    plt.xticks(rotation=label_rotation)
+
+    return fig if return_fig else None
+
+
+def _dendrogram_obsolete(
```

Comment:
Should we delete the obsoleted function here?

Reply:
Yes. When this PR is ready to merge, I'll delete it. I keep it here just for testing purpose.